### PR TITLE
restore tipannotations; fix #53

### LINKED ISSUE #53: Plot tree doesn't show the tips
Hi! 

We have just found a problem with the Plot recipe for Phylo's trees. The plots do not show the tips, as shown in https://github.com/EcoJulia/NCBITaxonomy.jl/issues/32. The code here can be used as an example of the problem: http://docs.ecojulia.org/NCBITaxonomy.jl/dev/phylo/ The misbehaviour can be related to changes in Plots. 

I really appreciate any help you can provide.

### PR DESCRIPTION
**Pull request by BenjaminDoran**
This should fix issue #53,  

### what I had to do
There may be a better way. but I put in a new `@series begin ... end` block with changes below, which seems to show the annotations again.

It required adding Plots as a dependency to get access to the Plots.text function. Pending issue https://github.com/JuliaPlots/RecipesBase.jl/issues/72 this seemed to be the only way to do it.

#### dendrogram example

```julia
if dend.showtips
    @series begin
        seriestype := :scatter
        markersize := 0
        series_annotations := map(x -> text(x[3], :left, dend.tipfont...), dend.tipannotations)
        getindex.(dend.tipannotations, 1), getindex.(dend.tipannotations, 2)
    end
end
```

#### fan example

```julia
if fan.showtips
    @series begin
        seriestype := :scatter
        markersize := 0
        series_annotations := map(x -> text(x[3], :left, rad2deg(adjust(x[2])), fan.tipfont...), fan.tipannotations)
        xys = map(x-> (_tocirc.(x[1], adjust.(x[2]))), fan.tipannotations)
        first.(xys), last.(xys)
    end
end
```

### Minimum example
```julia
using Phylo
using StatsPlots
tree = rand(Ultrametric(10))
plot(tree, treetype=:dendrogram, size=(400,400))
tree = rand(Ultrametric(10))
plot(tree, treetype=:fan, size=(400,400))
```

image

image

### CODE REVIEW

**Comment by mkborregaard** [comment]
Great, I'll have a look. I can say immediately that adding a Plots dep is not acceptable, but I don't think that should be necessary either.

File: src/plot.jl [hunk 1]
```diff
@@ -85,7 +86,14 @@ struct Fan; x; y; tipannotations; marker_x; marker_y; showtips; tipfont; marker_
             end
         end
     end
-    dend.showtips && (annotations := map(x -> (x[1], x[2], (x[3], :left, dend.tipfont...)), dend.tipannotations))
+    if dend.showtips
+        @series begin
+            seriestype := :scatter
+            markersize := 0
+            series_annotations := map(x -> text(x[3], :left, dend.tipfont...), dend.tipannotations)
```
Comment (mkborregaard) [inline src/plot.jl]: do you need the `text` here?

**Comment by mkborregaard** [comment]
I can't work out why the original code doesn't work as it is. Consider
```julia
struct MyType end

@recipe function f(x::MyType)
    seriestype := :myseries
    1:10, 1:10
end

@recipe function f(::Type{Val{:myseries}}, x, y, z)
    annotations := [(2, 4, ( "test", :right, 20)), (3, 4, ("test2", :left, 10))]
    seriestype := :path
    x, y
end

plot(MyType())
```
Skærmbillede 2021-07-06 kl  15 43 35

Seems to me like exactly what we have in the original code?

File: src/plot.jl (hunk 1 above)
Comment (BenjaminDoran) [reply src/plot.jl]: for this solution, yes. It writes out the unformatted entire tuple otherwise.

**Comment by BenjaminDoran** [comment]
Thanks for the test example, I don't really know why, but changing the location of the `annotations` argument fixes the issue.

i.e. 

#### Doesn't work
```julia
struct MyType end
struct myseries; x; y; end

@recipe function f(x::MyType)
    seriestype := :myseries
    myseries(1:10, 1:10)
end


@recipe function f(ms::myseries)
    @series begin
        seriestype := :path
        ms.x, ms.y
    end
    annotations := [(2, 4, ( "test", :right, 20)), (3, 4, ("test2", :left, 10))]
    nothing
end

plot(MyType())
```
#### works
```julia
struct MyType end
struct myseries; x; y; end

@recipe function f(x::MyType)
    seriestype := :myseries
    myseries(1:10, 1:10)
end


@recipe function f(ms::myseries)
    annotations := [(2, 4, ( "test", :right, 20)), (3, 4, ("test2", :left, 10))]
    @series begin
        seriestype := :path
        ms.x, ms.y
    end
    nothing
end

plot(MyType())
```
I agree it is a much better solution, than needing the plots dependency. I reverted my commit and just moved your code blocks to above the other series blocks. And, it seems to work now!

image

File: src/plot.jl [hunk 2]
```diff
@@ -86,14 +88,6 @@ struct Fan; x; y; tipannotations; marker_x; marker_y; showtips; tipfont; marker_
...
-    if dend.showtips
-        @series begin
-            seriestype := :scatter
-            markersize := 0
-            series_annotations := map(x -> text(x[3], :left, dend.tipfont...), dend.tipannotations)
-            getindex.(dend.tipannotations, 1), getindex.(dend.tipannotations, 2)
-        end
-    end
     primary = false
     label = ""
```
Comment (BenjaminDoran) [inline src/plot.jl]: also what are these two lines doing? are they meant to be ``primary := false; label := ""``? or just deleted?

Comment (mkborregaard) [reply src/plot.jl]: `primary := false; label := ""` as you say :+1:

Comment (mkborregaard) [reply src/plot.jl]: do you mind updating that before I merge?

**Comment by mkborregaard** [comment]
Awesome!

**Comment by mkborregaard** [comment]
Thanks a lot @BenjaminDoran !

**Comment by richardreeve** [comment]
Yes, thanks so much for fixing this - knowing nothing about plotting it has just been sitting around mildly annoying me for ages! Looking at the PR, it's not remotely clear to me why this solution works either, but never mind :)

**Pull request merged**